### PR TITLE
Static functions for uploading standalone data

### PR DESCRIPTION
**Pull request by ronichoudhury**
Includes a pair of functions for uploading free-standing data into a new file in
an Item or Folder.

### CODE REVIEW

**Comment by ronichoudhury** [comment]
I am not sure how/where to write tests for this.  Can someone point me the right way?

**Comment by zachmullen** [comment]
So, since this still depends on girder.app.js, I think we should turn these into model methods of `FileModel` and make them flexible in two ways:

1. The first argument to `uploadToFoo` should be either a string (the id) or a model object of that type, e.g.

```
uploadToFolder: function (modelOrId, data) {
    if (typeof(modelOrId) === 'string') {
        modelOrId = new girder.models.FolderModel({_id: modelOrId});
    }
    this.upload(modelOrId, data);
}
```

2. The second argument should be allowed to be a `Blob` like we support now, as well as a string of already-in-memory data, e.g.

```
upload: function (model, data) {
    if (typeof(data) === 'string') {
        data = new Blob([data]);
    }
   ...
}
```

Then we should make the [cookbook section](http://girder.readthedocs.org/en/latest/developer-cookbook.html#upload-a-file) about this a lot better by describing these options.

**Comment by ronichoudhury** [comment]
Sounds good.  What do we do about uploading to folders vs. items?

**Comment by zachmullen** [comment]
```
uploadToFolder: function (modelOrId, data) {
    if (typeof(modelOrId) === 'string') {
        modelOrId = new girder.models.FolderModel({_id: modelOrId});
    }
    this.upload(modelOrId, data);
},

uploadToItem: function (modelOrId, data) {
    if (typeof(modelOrId) === 'string') {
        modelOrId = new girder.models.ItemModel({_id: modelOrId});
    }
    this.upload(modelOrId, data);
}
```

Just two functions like you initially suggested. I think that's a pretty grokable API, but if someone can think of a better way I'm all ears.

**Comment by zachmullen** [comment]
Feel free to still use underscore's handy partial function application to reduce duplication from my example.

**Comment by ronichoudhury** [comment]
I do love to slip in functional programming whenever no one's looking :neckbeard:

Thanks for the guidance.  I'll merge these ideas into the models in a bit.

**Comment by ronichoudhury** [comment]
Quick question:  if a ``FileModel`` already has an ``_id`` (i.e., isn't "new" in the Backbone sense), what would happen when you call ``uploadToFolder`` or ``uploadToItem`` on it?

I ask because eventually I will also want a utility function that can do contents replacement on an existing file as well.

**Comment by ronichoudhury** [comment]
In answer to my last question, I would create a ``FileModel`` with a given id, then call ``updateContents()`` on it, but as part of this effort we'll want to generalize the ``file`` parameter of that function as well.  Is that right?

**Comment by zachmullen** [comment]
That's right :)

**Comment by ronichoudhury** [comment]
@zachmullen let me know if this is the right idea.

**Comment by zachmullen** [comment]
Looks great

**Comment by ronichoudhury** [comment]
What should I do to move this forward?

**Comment by zachmullen** [comment]
We just need a test that calls this with a string. We have existing test utilities that perform uploads, see https://github.com/girder/girder/blob/master/clients/web/test/testUtils.js#L703

We can add a new part to the `dataSpec.js` that uploads another item using the string argument.

**Comment by zachmullen** [comment]
And we could tweak something in the test code or the application code to make sure we call either `uploadToItem` or `uploadToFolder` instead of just `upload`, so that we can cover the new wrapping methods.

**Comment by ronichoudhury** [comment]
Thanks for the pointers.  I'll take a look soon when some of my immediate fires are out.

**Comment by ronichoudhury** [comment]
@zachmullen let me know if the changes to the S3 upload handler look right.

**Comment by zachmullen** [comment]
So I was wrong, I just tested a multichunk s3 upload, it is simply

    handler.trigger('g:upload.complete', resp);

The XHR is the request after that to the S3 service, the resp is the file response from Girder.

**Comment by ronichoudhury** [comment]
Do we need to ``delete resp.s3FinalizeRequest`` or anything like that?

**Comment by zachmullen** [comment]
Yeah probably makes sense to remove that.

File: clients/web/src/model.js [hunk 1]
```diff
@@ -4,6 +4,7 @@
  */
 girder.Model = Backbone.Model.extend({
     resourceName: null,
+    idAttribute: "_id",
```
Comment (zachmullen) [inline clients/web/src/model.js]: Crockford quotes

**Comment by zachmullen** [comment]
I'm getting the same failure locally that travis is showing, looks like a legit failure.

**Comment by ronichoudhury** [comment]
Is it an S3 error?  I'm not sure I am equipped to diagnose/fix that :(

**Comment by zachmullen** [comment]
I will investigate tonight.

**Comment by zachmullen** [comment]
Can you push this topic to girder/girder and reopen the PR? I will probably make a commit onto the branch to fix that test.

**Comment by zachmullen** [comment]
Closing in favor of https://github.com/girder/girder/pull/755

**Pull request closed without merging**